File: apps/server/src/application/shopify/handlers/shopify-sync.handler.ts
// src/application/shopify/handlers/shopify-sync.handler.ts

import { Option } from 'effect'
import {
  annotateLogs,
  gen,
  logInfo,
  logWarning,
  mapError,
  withLogSpan,
  type Effect,
} from 'effect/Effect'

import {
  markSynced,
  ProductStatus,
  type PilotProduct,
  type PilotProductPublished,
  type PilotProductUpdated,
  type ProductId,
  type ShopifyProductId,
} from '../../../domain/pilot'
import {
  Clock,
  MessageHandlerError,
  PilotProductRepository,
  ShopifyClient,
  type MessageHandler,
} from '../../../ports/driven'

// ============================================
// SYNC EVENT TYPE
// Both events have the same structure for sync
// ============================================

export type ShopifySyncEvent = PilotProductPublished | PilotProductUpdated

/**
 * Handles PilotProductPublished and PilotProductUpdated events to sync products to Shopify
 *
 * This handler:
 * - Listens to PilotProductPublished and PilotProductUpdated events from RabbitMQ
 * - Checks product status:
 *   - DRAFT: Skip sync (work in progress)
 *   - PUBLISHED: Sync to Shopify (create or update)
 *   - ARCHIVED: Archive on Shopify
 * - Updates PilotProduct syncStatus with Shopify ID
 * - On failure after max retries, marks syncStatus as SyncFailed
 */
export const shopifySyncHandler: MessageHandler<
  ShopifySyncEvent,
  PilotProductRepository | ShopifyClient | Clock
> = (event) =>
  gen(function* () {
    const { product, productId, correlationId, userId } = event

    yield* logInfo('Processing product for Shopify sync').pipe(
      annotateLogs({
        productId,
        correlationId,
        userId,
        status: product.status,
        eventType: event._tag,
      }),
      withLogSpan('shopify-sync.process')
    )

    switch (product.status) {
      case ProductStatus.DRAFT:
        yield* logInfo('Product is DRAFT, skipping Shopify sync').pipe(annotateLogs({ productId }))
        return

      case ProductStatus.PUBLISHED:
        yield* syncToShopify(event, product)
        return

      case ProductStatus.ARCHIVED:
        yield* archiveOnShopify(event, product)
        return
    }
  })

// ============================================
// SYNC TO SHOPIFY (PUBLISHED status)
// ============================================

const syncToShopify = (
  event: ShopifySyncEvent,
  product: PilotProduct
): Effect<void, MessageHandlerError, PilotProductRepository | ShopifyClient | Clock> =>
  gen(function* () {
    const { productId } = event
    const shopifyClient = yield* ShopifyClient

    const shopifyProductId = yield* shopifyClient
      .syncProduct(product)
      .pipe(mapError((error) => new MessageHandlerError({ event, cause: error })))

    yield* logInfo('Product synced to Shopify').pipe(annotateLogs({ shopifyProductId }))

    yield* updateSyncStatus(event, productId, shopifyProductId)
  })

// ============================================
// ARCHIVE ON SHOPIFY (ARCHIVED status)
// ============================================

const archiveOnShopify = (
  event: ShopifySyncEvent,
  product: PilotProduct
): Effect<void, MessageHandlerError, ShopifyClient> =>
  gen(function* () {
    const { productId } = event

    if (product.syncStatus._tag !== 'Synced') {
      yield* logInfo('Product not synced to Shopify, nothing to archive').pipe(
        annotateLogs({ productId, syncStatus: product.syncStatus._tag })
      )
      return
    }

    const { shopifyProductId } = product.syncStatus

    yield* logInfo('Archiving product on Shopify').pipe(
      annotateLogs({ productId, shopifyProductId })
    )

    const shopifyClient = yield* ShopifyClient

    yield* shopifyClient
      .archiveProduct(shopifyProductId)
      .pipe(mapError((error) => new MessageHandlerError({ event, cause: error })))

    yield* logInfo('Product archived on Shopify').pipe(
      annotateLogs({ productId, shopifyProductId })
    )
  })

// ============================================
// UPDATE SYNC STATUS
// ============================================

const updateSyncStatus = (
  event: ShopifySyncEvent,
  productId: ProductId,
  shopifyProductId: ShopifyProductId
): Effect<void, MessageHandlerError, PilotProductRepository | Clock> =>
  gen(function* () {
    const pilotRepo = yield* PilotProductRepository
    const clock = yield* Clock
    const now = yield* clock.now()

    const existingProduct = yield* pilotRepo
      .findById(productId)
      .pipe(mapError((error) => new MessageHandlerError({ event, cause: error })))

    if (Option.isNone(existingProduct)) {
      yield* logWarning('Product not found in repository, skipping syncStatus update').pipe(
        annotateLogs({ productId })
      )
      return
    }

    const currentProduct = existingProduct.value

    const updatedProduct = markSynced(currentProduct, shopifyProductId, now)

    yield* pilotRepo
      .update(updatedProduct)
      .pipe(mapError((error) => new MessageHandlerError({ event, cause: error })))

    yield* logInfo('Updated product syncStatus to Synced').pipe(
      annotateLogs({ productId, shopifyProductId, syncedAt: now.toISOString() })
    )
  })

File: apps/server-java/src/main/java/com/maisonamane/application/shopify/handler/ShopifySyncHandler.java
package com.maisonamane.application.shopify.handler;

import com.maisonamane.domain.pilot.aggregate.PilotProduct;
import com.maisonamane.domain.pilot.valueobject.ShopifyProductId;
import com.maisonamane.domain.pilot.valueobject.SyncStatus;
import com.maisonamane.port.driven.error.PersistenceError;
import com.maisonamane.port.driven.repository.PilotProductRepository;
import com.maisonamane.port.driven.service.Clock;
import com.maisonamane.port.driven.service.ShopifyClient;
import com.maisonamane.port.driven.service.ShopifyClient.ShopifyError;
import io.vavr.control.Either;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class ShopifySyncHandler {

    private static final Logger log = LoggerFactory.getLogger(ShopifySyncHandler.class);

    private final ShopifyClient shopifyClient;
    private final PilotProductRepository repository;
    private final Clock clock;

    public ShopifySyncHandler(
            ShopifyClient shopifyClient, PilotProductRepository repository, Clock clock) {
        this.shopifyClient = shopifyClient;
        this.repository = repository;
        this.clock = clock;
    }

    public sealed interface SyncError {
        String message();

        record ShopifyFailure(String message, ShopifyError cause) implements SyncError {}

        record PersistenceFailure(String message, PersistenceError cause) implements SyncError {}

        record AlreadySynced(String message) implements SyncError {}
    }

    public Either<SyncError, PilotProduct> handle(PilotProduct product, String correlationId) {

        log.info(
                "Syncing product to Shopify - correlationId: {}, productId: {}",
                correlationId,
                product.id());

        // Check if already synced
        if (product.syncStatus().isSynced()) {
            log.info(
                    "Product already synced - correlationId: {}, productId: {}",
                    correlationId,
                    product.id());
            return Either.left(
                    new SyncError.AlreadySynced("Product " + product.id() + " is already synced"));
        }

        return shopifyClient
                .syncProduct(product)
                .mapLeft(e -> (SyncError) new SyncError.ShopifyFailure("Shopify sync failed", e))
                .flatMap(shopifyId -> updateSyncStatus(product, shopifyId, correlationId));
    }

    private Either<SyncError, PilotProduct> updateSyncStatus(
            PilotProduct product, ShopifyProductId shopifyId, String correlationId) {

        var now = clock.now();
        var synced = SyncStatus.synced(shopifyId, now);
        var updatedProduct = product.withSyncStatus(synced, now);

        return repository
                .update(updatedProduct)
                .peek(
                        p ->
                                log.info(
                                        "Product sync status updated - correlationId: {}, productId: {}, shopifyId: {}",
                                        correlationId,
                                        p.id(),
                                        shopifyId))
                .mapLeft(e -> new SyncError.PersistenceFailure("Failed to update sync status", e));
    }
}
